### PR TITLE
Migrating some of the authenticated middleware into utilities and plugins

### PR DESCRIPTION
**Pull request by codyrancher**
This is a series of smaller PRs where the end goal is to get to get to https://github.com/rancher/dashboard/pull/10525 where we remove all nuxt middleware including authenticated.

### CODE REVIEW

File: cypress/e2e/tests/setup/rancher-setup.spec.ts [hunk 1]
```diff
@@ -8,7 +8,7 @@ describe('Rancher setup', { tags: ['@adminUserSetup', '@standardUserSetup', '@se
     cy.visit('');
 
     new RancherSetupPagePo().hasInfoMessage();
-    cy.url().should('equal', `${ Cypress.config().baseUrl }/auth/login`);
```
Comment (codyrancher) [inline cypress/e2e/tests/setup/rancher-setup.spec.ts:11]: This appeared to have a race-condition which was being exacerbated by these changes causing this test to be more flaky. Depending on when the url is checked it may contain the timed-out query param.

File: shell/middleware/authenticated.js [hunk 2]
```diff
@@ -239,26 +84,17 @@ export default async function({
...
 
       next();
     });
 
     // Call it for the initial pageload
-    const redirected = setProduct(store, route, redirect);
-
-    if (redirected) {
-      return redirected();
+    if (await store.dispatch('setProduct', route)) {
```
Comment (codyrancher) [inline shell/middleware/authenticated.js:95]: We don't return a redirect method anymore but we still need to exit from this middleware method

File: cypress/e2e/tests/setup/rancher-setup.spec.ts (hunk 1 above)
Comment (cnotv) [reply cypress/e2e/tests/setup/rancher-setup.spec.ts:11]: I guess there are query paramters added or?

File: shell/middleware/authenticated.js (hunk 2 above)
Comment (cnotv) [reply shell/middleware/authenticated.js:95]: In the other returns it's used `false`, should we keep the same?

File: shell/plugins/back-button.js [hunk 3]
```diff
@@ -1,3 +1,12 @@
-window.addEventListener('popstate', () => {
-  window._popStateDetected = true;
-});
+export default function({ app }) {
+  window.addEventListener('popstate', () => {
+    window._popStateDetected = true;
+  });
+
+  app.router.afterEach((to, from) => {
```
Comment (cnotv) [inline shell/plugins/back-button.js:6]: Could we have router-related logic inside the router configuration?
Sparse code is quite hard to track.

File: shell/plugins/back-button.js [hunk 4]
```diff
@@ -1,3 +1,12 @@
...
+    // Clear state used to record if back button was used for navigation
+    setTimeout(() => {
```
Comment (cnotv) [inline shell/plugins/back-button.js:8]: Should this be a Vue `nextTick()`?

File: shell/middleware/authenticated.js [hunk 5]
```diff
@@ -1,210 +1,55 @@
...
-      return redirect({ name: 'auth-setup' });
-    }
+    await store.dispatch('auth/handleUserMustChangePassword');
 
     // In newer versions the API calls return the auth state instead of having to make a new call all the time.
     const fromHeader = store.getters['auth/fromHeader'];
 
     if ( fromHeader === 'none' ) {
-      noAuth();
+      store.dispatch('auth/noAuth');
```
Comment (cnotv) [inline shell/middleware/authenticated.js:36]: Could we turn this whole chain of conditions into a switch for clarity?

File: shell/store/auth.js [hunk 6]
```diff
@@ -130,10 +130,94 @@ export const actions = {
...
   },
 
+  async loggedIn({ dispatch, commit }) {
+    const me = await dispatch('findMe');
+
+    commit('hasAuth', true);
+    commit('loggedInAs', me.id);
+  },
+
+  notLoggedIn({ commit }) {
```
Comment (cnotv) [inline shell/store/auth.js:140]: I do not think we should handle redirects in the store.
The idea of the state management and actions is to commit changes to the store. If nothing is happening there, then it's either a helper or a service (if it contains its own variables) and it does not need to pass from the store.
The very fact that we pass the router logic as single actions and not value to a router logic set somewhere else makes it very complicated to find out where things happen.

File: shell/store/auth.js [hunk 7]
```diff
@@ -130,10 +130,94 @@ export const actions = {
...
+  // First thing we do in loadManagement is fetch principals anyway.... so don't ?me=true here
+    const principals = await dispatch('rancher/findAll', {
+      type: NORMAN.PRINCIPAL,
+      opt:  {
+        url:                  '/v3/principals',
+        redirectUnauthorized: false,
+      }
+    }, { root: true });
+
+    const me = findBy(principals, 'me', true);
```
Comment (cnotv) [inline shell/store/auth.js:203]: me? 😅

File: shell/store/index.js [hunk 8]
```diff
@@ -809,6 +812,131 @@ export const actions = {
...
+
+      const res = getters['management/byId'](MANAGEMENT.SETTING, SETTING.FIRST_LOGIN);
+      const plSetting = getters['management/byId'](MANAGEMENT.SETTING, SETTING.PL);
+
+      isFirstLogin = res?.value === 'true';
+
+      if (!defaultPassword && plSetting?.value === 'Harvester') {
+        defaultPassword = 'admin';
+      }
+    } catch (e) {
```
Comment (cnotv) [inline shell/store/index.js:908]: Why no handling?

File: shell/store/index.js [hunk 9]
```diff
@@ -809,6 +812,131 @@ export const actions = {
...
+        const plSetting = await dispatch('rancher/find', {
+          type: 'setting',
+          id:   SETTING.PL,
+          opt:  { url: `/v3/settings/${ SETTING.PL }` }
+        });
+
+        if (!defaultPassword && plSetting?.value === 'Harvester') {
+          defaultPassword = 'admin';
+        }
+      } catch (e) {
```
Comment (cnotv) [inline shell/store/index.js:930]: Also here, what should we do in this case if it fails?

**Review by cnotv — CHANGES_REQUESTED** [review]
I like how we cleaned up everything, definitely easier to read.

Before everything, are we testing all these changes here? If not, we should as a bare minimum create issues for handling tests of this whole logic.

Also, could we split this function into clearer actions/features, for `authenticated.js`?

File: cypress/e2e/tests/setup/rancher-setup.spec.ts (hunk 1 above)
Comment (codyrancher) [reply cypress/e2e/tests/setup/rancher-setup.spec.ts:11]: Correct ?timed-out.

image

File: shell/store/auth.js (hunk 6 above)
Comment (rak-phillip) [reply shell/store/auth.js:140]: I also share similar concerns about handling routing from the store. Unfortunately, this is already a pattern that is pervasive throughout dashboard. We explicitly expose the router to the store for this very purpose

https://github.com/rancher/dashboard/blob/d8c0cf69bf1be6229ab7e58ac4b49dc3ab07d55d/shell/initialize/index.js#L46-L47

My initial instinct is that functions that handle dispatching actions and redirects  (`setProduct`, `invalidResource`, `handleFirstLogin`, etc..) would better fit as utility functions, but I can't say that's a better abstraction than what's represented here..

@codyrancher have we considered alternatives to using the store for these functions?

**Review by rak-phillip — COMMENTED** [review]
I don't have much to add outside of what's already been mentioned. Handling redirects from within the store isn't my preferred approach and I'd like to better understand what approaches we have available to us.

File: shell/middleware/authenticated.js (hunk 5 above)
Comment (codyrancher) [reply shell/middleware/authenticated.js]: Moved to a switch

File: shell/store/auth.js (hunk 6 above)
Comment (codyrancher) [reply shell/store/auth.js]: I moved it to a utility.

File: shell/utils/authentication.js
Comment (codyrancher) [inline shell/utils/authentication.js:1]: I recognize that there's an auth.js but it's doing authorization and not authentication. I think that one should probably be renamed in a separate pr.

File: shell/middleware/authenticated.js (hunk 2 above)
Comment (codyrancher) [reply shell/middleware/authenticated.js:73]: I'm not sure what you're asking here. This should still be the same behavior given that setProduct doesn't return a redirect method.

File: shell/store/auth.js (hunk 7 above)
Comment (cnotv) [reply shell/store/auth.js]: hahah cmon 😄 
I guess also this case we were just moving the code.

File: shell/plugins/back-button.js (hunk 4 above)
Comment (cnotv) [reply shell/plugins/back-button.js:8]: Good idea

**Review by cnotv — APPROVED** [review]
LGTM

**Comment by codyrancher** [comment]
I decided I'm just going to abandon this PR. We'll be handling the middleware migration differently.
